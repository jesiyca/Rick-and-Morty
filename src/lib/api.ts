import { Character, CharactersResponse } from '@/types/character';

interface FetchCharactersParams {
    page?: number;
    name?: string;
    species?: string;
    status?: string;
    type?: string;
    gender?: string;
  }

export async function fetchCharacter(id: number): Promise<Character> {
    const res = await fetch(`https://rickandmortyapi.com/api/character/${id}`);
    if (!res.ok) throw new Error(`Failed to fetch character with id = ${id} with error = ${res.statusText}`);
    return res.json();
  }

export async function fetchCharacters(params: FetchCharactersParams = {}): Promise<CharactersResponse> {
    const query = new URLSearchParams();

    for (const [key, value] of Object.entries(params)) {
      if (value != null && value !== '') {
        query.append(key, String(value));
      }
    }
  
    const res = await fetch(`https://rickandmortyapi.com/api/character/?${query}`);
    if (!res.ok) {
      throw new Error(`Failed to fetch characters for ${query} with error = ${res.statusText}`);
    }
  
    return res.json();
  }