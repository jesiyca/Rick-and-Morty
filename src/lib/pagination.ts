/**
 * Generates a list of pages for pagination UI, with ellipsis (-1) placeholders.
 */
export function getPageRange(currentPage: number, totalPages: number, visiblePages = 5): number[] {
    const pages: number[] = [];
  
    if (totalPages <= visiblePages) {
      return Array.from({ length: totalPages }, (_, i) => i + 1);
    }
  
    const showLeftEllipsis = currentPage > 3;
    const showRightEllipsis = currentPage < totalPages - 2;
  
    pages.push(1);
  
    if (showLeftEllipsis) {
      pages.push(-1); // Left ellipsis
    }
  
    const start = Math.max(2, currentPage - 1);
    const end = Math.min(totalPages - 1, currentPage + 1);
  
    for (let i = start; i <= end; i++) {
      pages.push(i);
    }
  
    if (showRightEllipsis) {
      pages.push(-1); // Right ellipsis
    }
  
    pages.push(totalPages);
  
    return pages;
  }