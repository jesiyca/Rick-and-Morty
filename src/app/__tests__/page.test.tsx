import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import Home from '../page';
import { useSearchParams, useRouter, usePathname } from 'next/navigation';
import { fetchCharacters } from '@/lib/api';

jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
  usePathname: jest.fn(),
}));

jest.mock('@/lib/api', () => ({ fetchCharacters: jest.fn() }));

const mockPush = jest.fn();
const characters = {
  info: { pages: 3 },
  results: [
    { id: 1, name: 'Rick Sanchez' },
    { id: 2, name: 'Morty Smith' },
  ],
};

describe('Home Page', () => {
  beforeEach(() => {
    (useSearchParams as jest.Mock).mockReturnValue({ get: () => '1', toString: () => 'page=1' });
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });
    (usePathname as jest.Mock).mockReturnValue('/');
    (fetchCharacters as jest.Mock).mockResolvedValue(characters);
  });

  afterEach(jest.clearAllMocks);

  it('matches snapshot', () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });

  it('renders characters from API', async () => {
    render(<Home />);
    expect(await screen.findByText('Rick Sanchez')).toBeInTheDocument();
    expect(screen.getByText('Morty Smith')).toBeInTheDocument();
  });

  it('navigates to character page on click', async () => {
    render(<Home />);
    fireEvent.click(await screen.findByText('Rick Sanchez'));
    expect(mockPush).toHaveBeenCalledWith('/character/1');
  });
});
