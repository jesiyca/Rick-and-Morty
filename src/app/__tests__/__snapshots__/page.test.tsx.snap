// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page matches snapshot 1`] = `
<div>
  <div
    class="grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]"
  >
    <main
      class="flex flex-col gap-4 row-start-2 items-center sm:items-start"
    >
      <header>
        Rick and Morty Characters
      </header>
      <div
        class="relative w-full overflow-x-auto"
        data-slot="table-container"
      >
        <table
          class="w-full caption-bottom text-sm"
          data-slot="table"
        >
          <caption
            class="text-muted-foreground mt-4 text-sm"
            data-slot="table-caption"
          >
            A list of Rick and Morty characters.
          </caption>
          <thead
            class="[&_tr]:border-b"
            data-slot="table-header"
          >
            <tr
              class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
              data-slot="table-row"
            >
              <th
                class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[100px]"
                data-slot="table-head"
              >
                <b>
                  Name
                </b>
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
            data-slot="table-body"
          />
        </table>
      </div>
      <nav
        aria-label="pagination"
        class="mx-auto flex w-full justify-center"
        data-slot="pagination"
        role="navigation"
      >
        <ul
          class="flex flex-row items-center gap-1"
          data-slot="pagination-content"
        >
          <li
            data-slot="pagination-item"
          >
            <a
              aria-label="Go to previous page"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pl-2.5 pointer-events-none opacity-50"
              data-slot="pagination-link"
              href="/?"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-chevron-left"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m15 18-6-6 6-6"
                />
              </svg>
              <span
                class="hidden sm:block"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            data-slot="pagination-item"
          >
            <a
              aria-current="page"
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
              data-active="true"
              data-slot="pagination-link"
              href="/?"
            >
              1
            </a>
          </li>
          <li
            data-slot="pagination-item"
          >
            <a
              aria-label="Go to next page"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pr-2.5 pointer-events-none opacity-50"
              data-slot="pagination-link"
              href="/?"
            >
              <span
                class="hidden sm:block"
              >
                Next
              </span>
              <svg
                aria-hidden="true"
                class="lucide lucide-chevron-right"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m9 18 6-6-6-6"
                />
              </svg>
            </a>
          </li>
        </ul>
      </nav>
    </main>
  </div>
</div>
`;

exports[`Home page should match snapshot 1`] = `
<div>
  <div
    class="grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]"
  >
    <main
      class="flex flex-col gap-4 row-start-2 items-center sm:items-start"
    >
      <header>
        Rick and Morty Characters
      </header>
      <div
        class="relative w-full overflow-x-auto"
        data-slot="table-container"
      >
        <table
          class="w-full caption-bottom text-sm"
          data-slot="table"
        >
          <caption
            class="text-muted-foreground mt-4 text-sm"
            data-slot="table-caption"
          >
            A list of Rick and Morty characters.
          </caption>
          <thead
            class="[&_tr]:border-b"
            data-slot="table-header"
          >
            <tr
              class="hover:bg-muted/50 data-[state=selected]:bg-muted border-b transition-colors"
              data-slot="table-row"
            >
              <th
                class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[100px]"
                data-slot="table-head"
              >
                <b>
                  Name
                </b>
              </th>
            </tr>
          </thead>
          <tbody
            class="[&_tr:last-child]:border-0"
            data-slot="table-body"
          />
        </table>
      </div>
      <nav
        aria-label="pagination"
        class="mx-auto flex w-full justify-center"
        data-slot="pagination"
        role="navigation"
      >
        <ul
          class="flex flex-row items-center gap-1"
          data-slot="pagination-content"
        >
          <li
            data-slot="pagination-item"
          >
            <a
              aria-label="Go to previous page"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pl-2.5 pointer-events-none opacity-50"
              data-slot="pagination-link"
              href="/?"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-chevron-left"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m15 18-6-6 6-6"
                />
              </svg>
              <span
                class="hidden sm:block"
              >
                Previous
              </span>
            </a>
          </li>
          <li
            data-slot="pagination-item"
          >
            <a
              aria-current="page"
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9"
              data-active="true"
              data-slot="pagination-link"
              href="/?"
            >
              1
            </a>
          </li>
          <li
            data-slot="pagination-item"
          >
            <a
              aria-label="Go to next page"
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pr-2.5 pointer-events-none opacity-50"
              data-slot="pagination-link"
              href="/?"
            >
              <span
                class="hidden sm:block"
              >
                Next
              </span>
              <svg
                aria-hidden="true"
                class="lucide lucide-chevron-right"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m9 18 6-6-6-6"
                />
              </svg>
            </a>
          </li>
        </ul>
      </nav>
    </main>
  </div>
</div>
`;
